tikv/deadlock.go

```diff
@@ -0,0 +1,217 @@
+func (dt *DeadlockDetector) Start() {
+	go dt.sendReqLoop()
+}
+
+// sendReqLoop will send detection request to leader, stream connection will be rebuilt and
+// a new recv goroutine using the same stream client will be created
+func (dt *DeadlockDetector) sendReqLoop() {
+	var (
+		err        error
+		rebuildErr error
+		req        *deadlockPb.DeadlockRequest
+		leaderAddr string
+	)
+	for {
+		if dt.streamCli == nil {
+			leaderAddr, rebuildErr = dt.rebuildStreamClient()
+			if rebuildErr != nil {
+				log.Errorf("rebuild connection to first region failed, err=%v", rebuildErr)
+				time.Sleep(3 * time.Second)
+				continue
+			}
+			log.Infof("rebuild stream connection to leader peer success, leaderAddr=%v", leaderAddr)
+		}
+		req = <-dt.sendCh
+		err = dt.streamCli.Send(req)
+		if err != nil {
+			log.Warnf("send req to addr=%v failed, err=%v, invalid current stream and try to rebuild connection",
+				leaderAddr, err)
+			dt.streamCli = nil
+		}
+	}
+}
+
+// recvLoop tries to recv response(current only deadlock error) from leader, break loop if errors happen
+func (dt *DeadlockDetector) recvLoop(streamCli deadlockPb.Deadlock_DetectClient) {
+	var (
+		err  error
+		resp *deadlockPb.DeadlockResponse
+	)
+	for {
+		resp, err = streamCli.Recv()
+		if err != nil {
+			log.Warnf("recv from failed, err=%v, stop receive", err)
+			break
+		}
+		// here only detection request will get response from leader
+		dt.waitMgr.WakeUpForDeadlock(resp)
+	}
+}
+
+func (dt *DeadlockDetector) sendReqToLeader(req *deadlockPb.DeadlockRequest) {
+	dt.sendCh <- req
+}
+
+func (dt *DeadlockDetector) handleRemoteTask(requestType deadlockPb.DeadlockRequestType,
+	txnTs uint64, waitForTxnTs uint64, keyHash uint64) {
+	detectReq := &deadlockPb.DeadlockRequest{}
+	detectReq.Tp = requestType
+	detectReq.Entry.Txn = txnTs
+	detectReq.Entry.WaitForTxn = waitForTxnTs
+	detectReq.Entry.KeyHash = keyHash
+	dt.sendReqToLeader(detectReq)
+}
+
+func (dt *DeadlockDetector) isLeader() bool {
+	return atomic.LoadInt32(&dt.role) == Leader
+}
+
+func (dt *DeadlockDetector) ChangeRole(newRole int32) {
+	atomic.StoreInt32(&dt.role, newRole)
+}
+
+// user interfaces
+// Cleanup processes cleaup task on local detector
+func (dt *DeadlockDetector) CleanUp(startTs uint64) {
+	if dt.isLeader() {
+		dt.detector.CleanUp(startTs)
+	} else {
+		dt.handleRemoteTask(deadlockPb.DeadlockRequestType_CleanUp, startTs, 0, 0)
+	}
+}
+
+// CleanUpWaitFor cleans up the specific wait edge in detector's wait map
+func (dt *DeadlockDetector) CleanUpWaitFor(txnTs, waitForTxn, keyHash uint64) {
+	if dt.isLeader() {
+		dt.detector.CleanUpWaitFor(txnTs, waitForTxn, keyHash)
+	} else {
+		dt.handleRemoteTask(deadlockPb.DeadlockRequestType_CleanUpWaitFor, txnTs, waitForTxn, keyHash)
+	}
+}
+
+// Detect will process the detection request on local detector
+func (dt *DeadlockDetector) Detect(txnTs uint64, waitForTxnTs uint64, keyHash uint64) error {
```

Comment:
Remove the local detector branch would make the code much simpler.

Reply:
This is used mainly for `Server.Detect`  RPC service handler

Reply:
We can separate to `DetectorClient` and `DetectorServer`, DetectorClient doesn't need to contain the `Detector` type.

Reply:
`DeadlockDetector` ->(split into)  `DetectorClient`  and `DetectorServer` ,  and 
`DetectorServer`  wraps the `detector` type and do real local detection(used by rpc handler),
`DetectorClient` warps connection to the leader node and is responcible for `send` and `recv`
do i understand right?

Reply:
yes

Reply:
Roger, refactored

Reply:
@coocood   only detection client interface kept, others removed